packages/flutter/lib/src/widgets/focus_manager.dart

```diff
@@ -68,330 +475,366 @@ class FocusNode extends ChangeNotifier {
+  @mustCallSuper
+  void _removeChild(FocusNode node) {
+    assert(_children.contains(node), "Tried to remove a node that wasn't a child.");
+    assert(node._parent == this);
+    assert(node._manager == _manager);
+
+    // If the child was (or requested to be) the primary focus, then unfocus it
+    // and cancel any outstanding request to be focused.
+    node.unfocus();
+
+    node._parent = null;
+    _children.remove(node);
+    assert(_manager == null || !_manager.rootScope.descendants.contains(node));
+  }
+
+  void _updateManager(FocusManager manager) {
+    _manager = manager;
+    for (FocusNode descendant in descendants) {
+      descendant._manager = manager;
+    }
+  }
+
+  // Used by FocusAttachment.reparent to perform the actual parenting operation.
+  @mustCallSuper
+  void _reparent(FocusNode child) {
+    assert(child != null);
+    assert(child != this, 'Tried to make a child into a parent of itself.');
+    if (child._parent == this) {
+      assert(_children.contains(child), "Found a node that says it's a child, but doesn't appear in the child list.");
+      // The child is already a child of this parent.
+      return;
+    }
+    assert(_manager == null || child != _manager.rootScope, "Reparenting the root node isn't allowed.");
+    assert(!ancestors.contains(child), 'The supplied child is already an ancestor of this node. Loops are not allowed.');
+    FocusNode oldPrimaryFocus;
+    if (child._manager != null) {
+      // We want to find out what the primary focus is, since the new child
+      // might be an ancestor of the primary focus, and it should move with the
+      // child.
+      oldPrimaryFocus = child.hasFocus ? child._manager._currentFocus : null;
+      assert(oldPrimaryFocus == null || oldPrimaryFocus == child || oldPrimaryFocus.ancestors.contains(child),
+        "child has focus, but primary focus isn't a descendant of it for some reason.");
+    }
+    // The child currently has focus, so we have to do some extra work to keep
+    // that focus, and to notify any scopes that used to be ancestors, and no
+    // longer have focus after we move it.
+    final Set<FocusNode> oldFocusPath = oldPrimaryFocus?.ancestors?.toSet() ?? <FocusNode>{};
+    child._parent?._removeChild(child);
+    _children.add(child);
+    child._parent = this;
+    child._updateManager(_manager);
+    if (oldPrimaryFocus != null) {
+      final Set<FocusNode> newFocusPath = _manager?._currentFocus?.ancestors?.toSet() ?? <FocusNode>{};
+      // Nodes that will no longer be focused need to be marked dirty.
+      for (FocusNode node in oldFocusPath.difference(newFocusPath)) {
+        node._markAsDirty();
+      }
+      // If the node used to have focus, make sure it keeps it's old primary
+      // focus when it moves.
+      oldPrimaryFocus.requestFocus();
+    }
+  }
+
+  /// Called by the _host_ [StatefulWidget] to attach a [FocusNode] to the focus tree.
```

Comment:
In other places this was called "attaching it to the element tree". Should we be consistent with the terminology?